lib/bitcoin/PsuedoRandomSampler.ts

```diff
@@ -0,0 +1,232 @@
+        n = n - d;
+        result = 1;
+      } else {
+        result = 0;
+      }
+
+    } while (this.psuedoRandomBitStream.getNextBit() === 0);
+
+    return result;
+  }
+
+  /**
+   * Get a number uniformly at random between 0 and n-1.
+   * This implementation uses the algorithm described in
+   * https://arxiv.org/pdf/1304.1916.pdf
+   *
+   * To see how it works, first consider n that is a power
+   * of 2, say 32. The while loop picks 5 random bits and returns
+   * the value encoded by the 5 bits as the random number; clearly,
+   * any of the values 0...31 are equally likely to be returned as
+   * desired.
+   *
+   * Now consider a non-power-of-2 such as 31. We again pick 5 bits and
+   * if these 5 bits encode a value <= 30, we return that value; otherwise,
+   * we repeat the process.
+   */
+  public getRandomNumber (n: number): number {
+    let n2 = 1;
+    let s = 0;
+
+    while (true) {
+      n2 = n2 * 2;
+      s = s * 2 + this.psuedoRandomBitStream.getNextBit();
+
+      if (n2 >= n) {
+        if (s < n) {
+          return s;
+        } else {
+          n2 = n2 - n;
+          s = s - n;
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Implement the reservoir sampling technique which maintains a uniform random
+ * sample over a "stream" of unknown size (https://en.wikipedia.org/wiki/Reservoir_sampling)
+ *
+ * We can add any number of elements to be considered for sampling and at any point request
+ * a sample of preconfigured size over the elements that have been added so far.
+ */
+export class ReservoirSampler {
+
+  /**
+   * Psuedo-randome generator for various random numbers we need for sampling
+   */
+  private psuedoRandomGenerator: PsuedoRandomGenerator | undefined;
+
+  /**
+   * The current sample
+   */
+  private sample: string[];
+
+  /**
+   * Number of elements added so far.
+   */
+  private streamSize: number = 0;
+
+  public constructor (private sampleSize: number) {
+    this.psuedoRandomGenerator = undefined;
+    this.sample = new Array<string>(this.sampleSize);
+  }
+
+  /**
+   * Add a new element to be considered for future sampling
+   */
+  public addElement (element: string): void {
+
+    // If we have not reached our sampling limit, we can simply add
+    // the current element to the sample
+    if (this.streamSize < this.sampleSize) {
+      this.sample[this.streamSize] = element;
+      this.streamSize++;
+      return;
+    }
+
+    // We have a full sampleSize of samples at this point.
+    // This element is picked to be in the sample with probability 1/streamSize
+    this.streamSize++;
+    if (this.psuedoRandomGenerator!.getBernoulliSample(this.sampleSize, this.streamSize) === 1) {
+
+      // If this element is picked, we need to decide which element to evict at random
+      const randIndex = this.psuedoRandomGenerator!.getRandomNumber(this.sampleSize);
+
+      // evict the current element at randIndex and store element instead
+      this.sample[randIndex] = element;
```

Comment:
Design: what is the reason for the random stream (Bernoulli) selection/eviction strategy? Is it to save memory in case of "large sliding window"? If that's the case, I feel storing large number of transaction IDs and perform a static random selection at the end isn't a bad alternative, *if* it leads to easier to understand and/or less code.

Reply:
We discussed this in our previous meeting. I simplified the reservoir sampling to not use two psuedorandom calls. The sampling code is around 4 lines - I am not sure if the static sampling would be much simpler.